Publicize/schedule button

artpi opened the pull request:
Sole purpose for this PR is being able to schedule a share action and write to scheduled actions endpoint.
<img width="644" alt="zrzut ekranu 2017-05-30 o 12 37 51" src="https://cloud.githubusercontent.com/assets/3775068/26579670/0cb0f7a6-4535-11e7-8e96-0c49979eb236.png">

## Integration testing (does it work?)

1. Build with a flag ( .`ENABLE_FEATURES=publicize-scheduling make run` )
2. Get a site with premium+ and publicize accounts connected
3. Go to /posts
4. Click "Share" under a post
5. Click on calendar icon
6. Pick a date in the future
7. Click "schedule post"
8. After that, you should see "We will share your post on..."
9. Test if really we have written to /scheduled table via wpsh:
`gpr SELECT * FROM publicize_scheduled_actions;`

## Unit testing

This PR also introduces some reducer unit tests:
`npm run test-client client/state/sharing/publicize/publicize-actions/test`

a8ci18n commented on client/my-sites/post-share/index.jsx:
```diff
@@ -169,7 +205,7 @@ class PostShare extends Component {
 			disabled={ this.isButtonDisabled() }
 		>
-			{ translate( 'Share post' ) }
+			{ this.state.scheduledDate ? translate( 'Schedule post' ) : translate( 'Share post' ) }
```
Hi! I've found a possible matching string that has already been translated **26** times:
`translate( 'Scheduled Post' )` *ES Score: 9*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/publicize/schedule-button)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

artpi replied:
Almost close, but not really :)

sirbrillig commented on client/my-sites/post-share/index.jsx:
```diff
@@ -123,7 +140,26 @@ class PostShare extends Component {
+				connections.map( connection => connection.keyring_connection_ID ),
+				this.state.skipped
+			);
+			this.props.schedulePostShareAction(
```
If you add a `return` on this line (or after), we can avoid the `else` below, which I think would improve the readability of the function.

artpi replied:
By having code inside a block, I can clearly see with intendations where is the else block and what to expect.
By using returns instead of blocks I have to wonder where the previous block started and I have to notice return to grok the fact that someone decided not to use else.
It is not immidiately clear for me how the condition ends. I think pattern of using return instead of else does not improve readability when there are only 2 conditions.
I agree that this is a sensible pattern when we have multilple conditions, but I don't think `else` keyword is so inherently bad that we should get rid of it.

sirbrillig commented on client/my-sites/post-share/index.jsx:
```diff
@@ -123,7 +140,26 @@ class PostShare extends Component {
+				servicesToPublish,
+			);
+		} else {
+			this.props.sharePost( this.props.siteId, this.props.postId, this.state.skipped, this.state.message );
```
Since some of these props have been deconstructed into `postId`, `site_id` above, why not deconstruct all of them and just pass them raw here? Or at least we can use those two.

artpi replied:
Sure thing! Changing.

sirbrillig commented on client/state/selectors/get-scheduled-publicize-share-action-time.js:
```diff
@@ -0,0 +1,16 @@
+ * @param {Object} state Global state tree
+ * @param {Number} siteId Site ID
+ * @param {Number} postId Post ID
+ * @return {String} time when newly scheduled share action will be published
```
Looks like this could actually be `String` or `false`. Although we could change the `get` below to default to an empty string. Not sure which is preferable for a failure.

artpi replied:
I like `false`  better. Will update!

sirbrillig commented on client/state/sharing/publicize/publicize-actions/actions.js:
```diff
@@ -104,3 +107,37 @@ export function editPostShareAction( siteId, postId, actionId, message, share_da
+				error,
+				connections
+			} ) )
+			.then( data => dispatch( {
```
Minor, but `data` could be changed to `items` here so we could use the shorthand object property below.

artpi replied:
Good call!

sirbrillig commented on client/state/sharing/publicize/publicize-actions/reducer.js:
```diff
@@ -97,11 +106,23 @@ export const editingSharePostAction = createReducer( {}, {
 		( state, { siteId, postId, actionId } ) => updateDataForPost( true, state, siteId, postId, actionId ),
 } );
 
+export const schedulingSharePostActionStatus = createReducer( {}, {
```
These reducers could use some tests to show what they do. `updateDataForPost` is a little confusing because it does not say in the name what it does (what is the "Data" that's updated, and how is it updated?). Often when a function is hard to name, it might be a good idea to split it into smaller sub-functions.

artpi replied:
Yes, I agree that naming of this function is unfortunate and would benefit from renaming.
This however, is definitely out of the scope of this PR.

sirbrillig commented on client/state/sharing/publicize/publicize-actions/reducer.js:
```diff
@@ -53,6 +57,11 @@ export const scheduled = createReducer( {}, {
 		( state, { siteId, postId, item } ) => updateDataForPost( item, state, siteId, postId, item.ID ),
+	[ PUBLICIZE_SHARE_ACTION_SCHEDULE_SUCCESS ]:
+		( state, { siteId, postId, items } ) => {
+			items.forEach( item => state = updateDataForPost( item, state, siteId, postId, item.ID ) );
```
Usually usages of `forEach` can be done more clearly with `map` or `reduce`. Could we do this in a different way that doesn't mutate `state` for each iteration?

I think I understand why you did it this way, though, because `updateDataForPost` appears to make some assumptions about changing just one item. It might be that `updateDataForPost` needs to be simplified so it can be used more easily here. I'm happy to help do that if you can help me understand what this is supposed to do.

Adding some tests for these actions would also help clarify how this is supposed to work and verify that it does.

artpi replied:
Usually yes, but I don't think that is the case. Of course, we can spend time in refactoring `updateDataForPost` so that is 10% more understandable for a person familiar with functional approach, but I do not believe that is a good time investment and am also pretty convinced that over-use of functional approach leads to a code that has much steeper barrier to entry for people not familiar with it.

That being said, I agree with you that we should simplify whenever possible, but I cannot imagine a way of doing so here where end result is better then this and is not invasive into changing every selector in this file because of changing `updateDataForPost` signature.
 
Since, like many selector trees, this one is deeply nested,  `updateDataForPost` updates deeply nested property in the redux tree, under site / post subtrees. on top of that, since actions are indexed by IDs it allows for changing the one with ID that matches.

Since we get arrays not indexed the same way from API, any use of array functions (such as map) is not helpful since we need to reindex IDs first. Doing several passes , reindexing idrs and then using array functions defeats the reason to use them since that does not improve readability in my opinion.

artpi replied:
> Adding some tests for these actions would also help clarify how this is supposed to work and verify that it does.

Done!

artpi commented:
I know that the title says "WIP" but since the label is "needs review" I added some comments.

> That's why I added it - it's better to catch errors early :)

Also, since I'm new it would help me if the testing instructions were more detailed. 

> In the WIP the work is still in progress :). Now its ready and I updated the comment, I hope ots good now

> Lastly I think that maybe this PR could be split into two: one with the reducers and actions and one with the components that use them. That way the data manipulation can be tested separately from the UI changes. Does that sound reasonable?

That would make testing each part that much harder (that would rely on unit test, bu providing the whole feature we can actually make an integration test and smoke test), welcome discussion on the shape of redux, introduce inconsistencies between the 2 parts and slow us down in the implementation.

I dislike vertical slicing of projects ( into "layers" ) and favor "horizontal" slicing where each new PR is an evolution to the feature introduced in the first one.
By working on the pieces separately and attaching them all at the end we risk that pieces will not fit together. By attaching all pieces to some body of work at each time, we are sure that they will fit where they need to.

I agree that this PR is not small. But it was much smaller without unit tests.

mcsf commented on client/my-sites/post-share/index.jsx:
```diff
@@ -262,6 +301,31 @@ class PostShare extends Component {
+		if ( this.props.scheduledAt ) {
+			return (
+				<Notice status="is-success" onDismissClick={ this.dismiss }>
+					{ translate( 'We`ll share your post on %s.', {
```
I noticed this via a P2 screenshot. @artpi, could we use an actual apostrophe instead of the backtick?

```diff
-{ translate( 'We`ll share your post on %s.', {
+{ translate( "We'll share your post on %s.", {
```
